File: src/snake_game_client/serializer.py

from state import State
from snake import Snake
from apple import Apple
import json

class Serializer:
     #  {
    #         "n_snakes": 2,
    #         "n_apples": 5,
    #         "board_width": 10,
    #         "board_height": 10,
    #         "turn":10,
    #         "snakes": [
    #             {
    #                 "id": 0,
    #                 "head": [x, y],
    #                 "tail": [
    #                     [x1, y1],
    #                     [x2, y2],
    #                     ...
    #                 ],
    #                 "moves_history": [
    #                     [x1, y1],
    #                     [x2, y2],
    #                     ...
    #                 ]
    #             },
    #             ...
    #         ],
    #         "apples": [
    #             {
    #                 "position": [x, y]
    #             },
    #             ...
    #         ],
    #         "eliminated_snakes": [0, 1, ...]
    #     }

    @staticmethod
    def serialize_state(state: State) -> str:
        jsonState = {
            "n_snakes": state.n_snakes,
            "n_apples": state.n_apples,
            "board_width": state.board_width,
            "board_height": state.board_height,
            "turn": state.turn,
            "snakes": [],
            "apples": [],
            "eliminated_snakes": list(state.eliminated_snakes),
            "idx_prev_snake": state.idx_prev_snake,
            "whoose_prev_turn": "py"
        }

        for i, snake in enumerate(state.snakes):
            snake_json = {
                "id": i,
                "head": [snake.head[0], snake.head[1]],
                "tail": [[x[0], x[1]] for x in snake.tail],
                "moves_history": [[x[0], x[1]] for x in snake.moves_history]
            }
            jsonState["snakes"].append(snake_json)

        jsonState["apples"] = [[apple.position[0], apple.position[1]] for apple in state.apples]

        return json.dumps(jsonState)
    
    @staticmethod
    def deserialize_state(jsonState: str) -> State:
        # print(jsonState)
        try:
            stateJson = json.loads(jsonState)
            n_snakes = stateJson["n_snakes"]
            n_apples = stateJson["n_apples"]
            board_width = stateJson["board_width"]
            board_height = stateJson["board_height"]
            turn = stateJson["turn"]

            # deserialization
            idx_prev_snake = stateJson["idx_prev_snake"]
            whoose_prev_turn = stateJson["whoose_prev_turn"]

            snakes = []
            apples = []

            for i in range(n_snakes):
                snake_json = stateJson["snakes"][i]
                snake = Snake(snake_json["head"][0], snake_json["head"][1])
                snake.tail = snake_json["tail"]
                snake.moves_history = snake_json["moves_history"]
                snakes.append(snake)

            for apple_position in stateJson["apples"]:
                x, y = apple_position
                apple = Apple(x, y)
                apples.append(apple)

            eliminated_snakes = set(stateJson["eliminated_snakes"])

            state = State(n_snakes, n_apples, board_width, board_height)
            state.turn = turn
            state.snakes = snakes
            state.apples = apples
            state.eliminated_snakes = eliminated_snakes
            state.idx_prev_snake = idx_prev_snake
            state.whoose_prev_turn = whoose_prev_turn
            return state
        except Exception as e:
            print(e)

File: cpp/python_cpp_binding/snake_game/scripts/Serializer.cpp
#include "Serializer.h"



string Serializer::serialize_apple(Apple &apple) {
    return "[" + to_string(apple.position.first) + ", " + to_string(apple.position.second) + "]";
}


string Serializer::serialize_snake(Snake &snake) {
    string json = "\"head\": [" + to_string(snake.head.first) + ", " + to_string(snake.head.second) + "], \"tail\": [";
    for (size_t i = 0; i < snake.tail.size(); ++i) {
        json += "[" + to_string(snake.tail[i].first) + ", " + to_string(snake.tail[i].second) + "]";
        if (i < snake.tail.size() - 1) {
            json += ", ";
        }
    }
    json += "]";

    // MAYBE HISTORY SERIALIZATION WOULD MAKE SENSE LATER ON, right now it's a waste of memory
    // add moves history
    json += ", \"moves_history\": [";
    for (size_t i = 0; i < snake.moves_history.size(); ++i) {
        json += "[" + to_string(snake.moves_history[i].first) + ", " + to_string(snake.moves_history[i].second) + "]";
        if (i < snake.moves_history.size() - 1) {
            json += ", ";
        }
    }
    json += "]";

    return json;
}


string Serializer::serialize_state(State& state){

    string json = "{ \"n_snakes\": " + to_string(state.n_snakes) + ", \"n_apples\": " + to_string(state.n_apples) + ", \"board_width\": " + to_string(state.board_width) + ", \"board_height\": " + to_string(state.board_height) + ", \"turn\": " + to_string(state.turn) + ", \"snakes\": [";
    
    for (size_t i = 0; i < state.snakes.size(); ++i) {
        json += "{ \"id\": " + to_string(i) + ", ";
        json += serialize_snake(state.snakes[i]);
        json += "}";
        if (i < state.snakes.size() - 1) {
            json += ", ";
        }
    }
    
    json += "], \"apples\": [";
    
    for (size_t i = 0; i < state.apples.size(); ++i) {
        json += serialize_apple(state.apples[i]);
        if (i < state.apples.size() - 1) {
            json += ", ";
        }
    }
    
    json += "], \"eliminated_snakes\": [";
    
    for (const auto& snake_idx : state.eliminated_snakes) {
        json += to_string(snake_idx);
        if (snake_idx != *state.eliminated_snakes.rbegin()) {
            json += ", ";
        }
    }
    
    json += "],";    
    json += "\"idx_prev_snake\":" + to_string(state.idx_prev_snake) + ",";
    json += "\"whoose_prev_turn\": \"cpp\"";
    json += "}";
    
    return json;
}


State Serializer::deserialize_state(const string& serialized_state){
    nlohmann::json jsonData = nlohmann::json::parse(serialized_state);
    int n_snakes = jsonData["n_snakes"];
    int n_apples = jsonData["n_apples"];
    int board_width = jsonData["board_width"];
    int board_height = jsonData["board_height"];
    int turn = jsonData["turn"];
    int idx_prev_snake = jsonData["idx_prev_snake"];
    string whoose_prev_turn = jsonData["whoose_prev_turn"];
    vector<Snake> snakes;
    
    for (const auto& snake_data : jsonData["snakes"]) {
        int id = snake_data["id"];
        int head_x = snake_data["head"][0];
        int head_y = snake_data["head"][1];
        Snake snake(head_x, head_y);
        
        for (const auto& tail_segment : snake_data["tail"]) {
            int tail_x = tail_segment[0];
            int tail_y = tail_segment[1];
            snake.tail.push_back(make_pair(tail_x, tail_y));
        }
        
        // MAYBE HISTORY SERIALIZATION WOULD MAKE SENSE LATER ON, right now it's a waste of memory
        for (const auto& move : snake_data["moves_history"]) {
            int move_x = move[0];
            int move_y = move[1];
            snake.moves_history.push_back(make_pair(move_x, move_y));
        }
        
        snakes.push_back(snake);
    }
    vector<Apple> apples;
    for (const auto& apple_data : jsonData["apples"]) {
        int apple_x = apple_data[0];
        int apple_y = apple_data[1];
        Apple apple(apple_x, apple_y);
        apples.push_back(apple);
    }
    set<int> eliminated_snakes;
    for (const auto& snake_idx : jsonData["eliminated_snakes"]) {
        eliminated_snakes.insert(snake_idx.get<int>());
    }
    State state(n_snakes, n_apples, board_width, board_height);
    state.turn = turn;
    state.snakes = snakes;
    state.apples = apples;
    state.eliminated_snakes = eliminated_snakes;
    state.idx_prev_snake = idx_prev_snake;
    state.whoose_prev_turn = whoose_prev_turn;
    return state;
}
